src/digest/base.cr

```diff
@@ -89,4 +88,28 @@ abstract class Digest::Base
 
     Base64.strict_encode(hashsum)
   end
+
+  def update(data) : self
+    update data.to_slice
+  end
+
+  def final : Bytes
+    dst = Bytes.new digest_size
+    final dst
+  end
+
+  # Dups and finishes the digest.
+  def digest : Bytes
+    dup.final
```

Comment:
It might seem like a legitimate idea to decouple finalization from current state for re-usability. I'd like to see a realistic use case for that, though. And I don't think this should be the default behaviour.
The `#digest` looks like the regular means to (simply) get the digest result and shouldn't come with a heat allocation as side effect. IMO such a method should be more explicitly named, for example `intermediary_digest`.

Reply:
Well, it's a departure from ruby.  In ruby `.to_s` calls `.digest`.  I assume the dupping behavior is so `.to_s` will work regardless of when it's called.  If `digest` mutates state either `to_s` prints the wrong result or worse `to_s` changes the resulting real hash value and breaks something.

In Crystal `OpenSSL::Digest` has a similar `.to_s` method.  So either showing the `.hexdigest` in `.to_s` has to go or the behavior needs to stay the same.

I could track state and disallow calling `final` more than once, leaving it to the programmer to call `dup.final` if they want to hash more than once, although I don't see why (other than inspection).  Printing the hash in `OpenSSL::Digest` obviously has to go with a change like this.

Reply:
Deprecate `#digest` and `#hexdigest` and replace with `#final`?

@watzon can comment if he likes regarding `#hexdigest`, `#hexfinal` or whatever it should be called.  `hexfinal(dst)` could avoid one `String(digest_length)` allocation.  I would use it but I'm not sure about other people.  

There was someone on gitter a few days ago looking to do as few memory allocations as possible when reading from `STDIN`.  Maybe it would be used if provided.

Reply:
Removing `#digest` in favour of `#final` seems fine. Theoretically, we could consider renaming the behaviour to `#digest`, but `#final` is more expressive.

`hexfinal(dst)` would be equivalent to `final(dst); dst.hexstring`. I don't think we need a short cut for that.

Reply:
I haven't run the benchmarks but could this be worth optimizing?

Creates two heap objects.
```
final.hexstring
```

Creates one heap object.
```
def hexfinal
  dst = StaticArray(...)
  String.new dst
end
```

Creates no heap objects.
```
def hexfinal(dst)
  tmp = StaticArray(...)
  final tmp
  some_hex_func tmp dst
end
```